cobalt/common/libc/tz/parse_posix_tz.cc

```diff
@@ -0,0 +1,463 @@
+
+  if (auto parsed_value = StringViewToInt(sv.substr(0, digits_end))) {
+    return {{*parsed_value, digits_end}};
+  }
+
+  return std::nullopt;
+}
+
+// Parses an optional, colon-prefixed time component (minutes or seconds).
+// Returns `true` on success (if the component was parsed, or if it wasn't
+// present). Returns `false` on a parse error (e.g., ":" not followed by
+// digits). On success, it updates `sv`, `consumed`, and `target_component`.
+bool ParseOptionalTimeComponent(std::string_view& sv,
+                                size_t& consumed,
+                                int& target_component) {
+  if (sv.empty() || sv.front() != ':') {
+    return true;  // Success: component is optional and not present.
+  }
+
+  // We have a ':', so a number must follow.
+  if (auto int_res = ParseInteger(sv.substr(1))) {
+    target_component = int_res->value;
+    // Consume the ':' plus the digits.
+    size_t component_consumed = 1 + int_res->consumed;
+    sv.remove_prefix(component_consumed);
+    consumed += component_consumed;
+    return true;
+  }
+
+  // Parse error: ':' was not followed by valid digits.
+  return false;
+}
+
+// Parses a time offset string (e.g., "8", "+2", or "-1:30") into seconds.
+// Returns the offset in seconds and the number of characters consumed.
+std::optional<ParseResult<TimeOffset>> ParseTimeOffset(
+    std::string_view time_offset_string) {
+  if (time_offset_string.empty()) {
+    return std::nullopt;
+  }
+
+  bool is_negative = false;
+  size_t consumed = 0;
+
+  // Handle optional leading '+' or '-' sign.
+  if (time_offset_string.front() == '+') {
+    time_offset_string.remove_prefix(1);
+    consumed++;
+  } else if (time_offset_string.front() == '-') {
+    is_negative = true;
+    time_offset_string.remove_prefix(1);
+    consumed++;
+  }
+
+  // --- Parse Hours (mandatory) ---
+  int total_seconds = 0;
+  if (auto hours_res = ParseInteger(time_offset_string)) {
+    total_seconds += hours_res->value * kHour;
+    consumed += hours_res->consumed;
+    time_offset_string.remove_prefix(hours_res->consumed);
+  } else {
+    return std::nullopt;  // Must have hours
+  }
+
+  // --- Parse Minutes (optional) ---
+  int minutes = 0;
+  if (ParseOptionalTimeComponent(time_offset_string, consumed, minutes)) {
+    total_seconds += minutes * kMinute;
+    // --- Parse Seconds (optional, only if minutes were parsable) ---
+    int seconds = 0;
+    if (ParseOptionalTimeComponent(time_offset_string, consumed, seconds)) {
+      total_seconds += seconds;
+    }
+  }
+
+  return {{is_negative ? -total_seconds : total_seconds, consumed}};
+}
+
+// Parses a date rule string (e.g., "M3.2.0" or "J60") into a DateRule struct.
+// This function remains strict, as an invalid rule is not recoverable.
+std::optional<DateRule> ParseDateRule(std::string_view date_rule_string) {
+  if (date_rule_string.empty()) {
+    return std::nullopt;
+  }
+
+  DateRule parsed_date_rule;
+  switch (date_rule_string.front()) {
```

Comment:
This function would be more readable if each case called (and potentially directly returned) its own function for that case